### PR TITLE
Created the Popover component / feature to build advanced stuff

### PR DESCRIPTION
**Pull request by GuillaumeNachury**

### CODE REVIEW

File: demo/react/components/popover/default.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import { Placements, Popover } from 'LumX';
+
+/////////////////////////////
+
+interface IProps {
+    /**
+     * The theme to use to display this demo.
+     */
+}
+const demoAnchorStyle: React.CSSProperties = {
```
[the remark this answers is not in the document]
Comment (Flo5k5) [reply demo/react/components/popover/default.tsx]: JSDoc

File: demo/react/components/popover/controled.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+const demoPopperStyle: React.CSSProperties = {
+    backgroundColor: 'black',
+    borderRadius: '3px',
+    color: 'white',
+    fontSize: '10px',
+    padding: '5px',
+    width: '123px',
+};
+
+function createDemoAnchor(): React.ReactNode {
```
[the remark this answers is not in the document]
Comment (Flo5k5) [reply demo/react/components/popover/controled.tsx]: JsDoc

File: demo/react/components/popover/controled.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ * The demo for the default <Popover>s.
+ *
+ * @return {React.ReactElement} The demo component.
+ */
+// tslint:disable: jsx-no-lambda
+const DemoComponent: React.FC<IProps> = (): React.ReactElement => {
+    // tslint:disable-next-line: typedef
+    const [isTooltipDisplayed, setTooltipDisplayed] = useState(false);
+
+    function toggleTooltipDisplay(): void {
```
[the remark this answers is not in the document]
Comment (Flo5k5) [reply demo/react/components/popover/controled.tsx]: JsDoc

[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/controled.tsx]: ok

File: demo/react/components/popover/controled.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        setTooltipDisplayed(!isTooltipDisplayed);
+    }
+
+    return (
+        <div>
+            <Button
+                size={ButtonSizes.s}
+                emphasis={ButtonEmphasises.medium}
+                onClick={(): void => toggleTooltipDisplay()}
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/controled.tsx]: ok

File: demo/react/components/popover/default.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/default.tsx]: ok

File: demo/react/components/popover/default.tsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        #CFCFCF,
+        #CFCFCF 3px,
+        #FFFFFF 3px,
+        #FFFFFF 6px
+      )`,
+    marginLeft: '3px',
+    marginRight: '3px',
+};
+
+function createDemoAnchor(): React.ReactNode {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/default.tsx]: ok

File: demo/react/components/popover/default.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    return <div style={demoAnchorStyle}>{`This element will act as the anchor`}</div>;
+}
+
+function createPopper(): React.ReactNode {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/default.tsx]: ok

File: demo/react/components/popover/controled.tsx [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            >
+                Toggle visibility
+            </Button>
+            <div style={{ height: 200, display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/controled.tsx]: ok

File: demo/react/components/popover/controled.tsx [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+import { Button, ButtonEmphasises, ButtonSizes, Placements, Popover } from 'LumX';
+
+/////////////////////////////
+
+interface IProps {
+    /**
+     * The theme to use to display this demo.
+     */
+}
+const demoAnchorStyle: React.CSSProperties = {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/controled.tsx]: ok

File: demo/react/components/popover/controled.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/controled.tsx]: ok

File: demo/react/components/popover/controled.tsx [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * The demo for the default <Popover>s.
+ *
+ * @return {React.ReactElement} The demo component.
+ */
+// tslint:disable: jsx-no-lambda
+const DemoComponent: React.FC<IProps> = (): React.ReactElement => {
+    // tslint:disable-next-line: typedef
+    const [isTooltipDisplayed, setTooltipDisplayed] = useState(false);
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/controled.tsx]: comme on l'a évoqué lors du weekly du 19, typer les `useState()` n'a vraiment aucun intérêt

File: demo/react/components/popover/default.tsx [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * The demo for the default <Popover>s.
+ *
+ * @return {React.ReactElement} The demo component.
+ */
+// tslint:disable-next-line: typedef
+const DemoComponent: React.FC<IProps> = (): React.ReactElement => {
+    return (
+        <div style={{ height: 200, display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/default.tsx]: ok

File: demo/react/components/popover/matchParentWidth.tsx [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+import { Placements, Popover } from 'LumX';
+
+/////////////////////////////
+
+interface IProps {
+    /**
+     * The theme to use to display this demo.
+     */
+}
+const demoAnchorStyle: React.CSSProperties = {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/matchParentWidth.tsx]: ok

File: demo/react/components/popover/matchParentWidth.tsx [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+const demoPopperStyle: React.CSSProperties = {
+    backgroundColor: 'black',
+    borderRadius: '3px',
+    color: 'white',
+    fontSize: '10px',
+    padding: '5px',
+    width: '100%',
+};
+
+function createDemoAnchor(width: number): React.ReactNode {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/matchParentWidth.tsx]: ok

File: demo/react/components/popover/matchParentWidth.tsx [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    return <div style={{ ...demoAnchorStyle, width }}>{'This element will act as the anchor'}</div>;
+}
+
+function createPopper(): React.ReactNode {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/matchParentWidth.tsx]: ok

File: demo/react/components/popover/matchParentWidth.tsx [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * The demo for the default <Popover>s.
+ *
+ * @return {React.ReactElement} The demo component.
+ */
+// tslint:disable: jsx-no-lambda
+const DemoComponent: React.FC<IProps> = (): React.ReactElement => {
+    // tslint:disable-next-line: typedef
+    const [isTooltipDisplayed, setTooltipDisplayed] = useState(false);
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/matchParentWidth.tsx]: vu

File: demo/react/components/popover/matchParentWidth.tsx [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+    function toggleTooltipDisplay(newVisibleState: boolean): void {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/matchParentWidth.tsx]: ok

File: demo/react/components/popover/matchParentWidth.tsx [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+        setTooltipDisplayed(newVisibleState);
+    }
+
+    return (
+        <div onMouseOver={(): void => toggleTooltipDisplay(true)} onMouseOut={(): void => toggleTooltipDisplay(false)}>
+            <div style={{ height: 200, display: 'flex', alignItems: 'center', justifyContent: 'space-around' }}>
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/matchParentWidth.tsx]: ok

File: demo/react/components/popover/offsets.tsx [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+import { Placements, Popover, PopperOffsets } from 'LumX';
+
+/////////////////////////////
+
+interface IProps {
+    /**
+     * The theme to use to display this demo.
+     */
+}
+const demoAnchorStyle: React.CSSProperties = {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/offsets.tsx]: ok

File: demo/react/components/popover/offsets.tsx [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+        #CFCFCF,
+        #CFCFCF 3px,
+        #FFFFFF 3px,
+        #FFFFFF 6px
+      )`,
+    marginLeft: '3px',
+    marginRight: '3px',
+};
+
+function createDemoAnchor(): React.ReactNode {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/offsets.tsx]: ok

File: demo/react/components/popover/offsets.tsx [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+/**
+ * The demo for the default <Popover>s.
+ *
+ * @return {React.ReactElement} The demo component.
+ */
+// tslint:disable-next-line: typedef
+const DemoComponent: React.FC<IProps> = (): React.ReactElement => {
+    return (
+        <div style={{ height: 200, display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/offsets.tsx]: ok

File: demo/react/components/popover/placements.tsx [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+import { Placements, Popover } from 'LumX';
+
+/////////////////////////////
+
+interface IProps {
+    /**
+     * The theme to use to display this demo.
+     */
+}
+const demoAnchorStyle: React.CSSProperties = {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/placements.tsx]: ok

File: demo/react/components/popover/placements.tsx [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+const demoPopperStyle: React.CSSProperties = {
+    backgroundColor: 'black',
+    borderRadius: '3px',
+    color: 'white',
+    fontSize: '10px',
+    padding: '5px',
+    width: '266px',
+};
+
+function createDemoAnchor(): React.ReactNode {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/placements.tsx]: ok

File: demo/react/components/popover/realCase.tsx [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+        emphasis={ButtonEmphasises.medium}
+        color={theme === ButtonThemes.dark ? 'light' : undefined}
+        size={ButtonSizes.s}
+        theme={theme}
+    >
+        Follow
+    </Button>
+);
+
+const demoActions: string[] = [mdiPhone, mdiCellphone, mdiEmail, mdiGoogleHangouts, mdiSlack];
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/realCase.tsx]: ok

File: demo/react/components/popover/realCase.tsx [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+            orientation={Orientations.horizontal}
+            onMouseEnter={(): void => toggleCardDisplay(true)}
+            onMouseLeave={(): void => toggleCardDisplay(false)}
+            size={UserBlockSize.m}
+        />
+    );
+
+    const popper: React.ReactNode = (
+        <div
+            style={{
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/realCase.tsx]: ok

File: demo/react/components/popover/realCase.tsx [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+                orientation={Orientations.vertical}
+                simpleAction={createSimpleAction(theme)}
+                multipleActions={createMultipleActions(theme)}
+            />
+        </div>
+    );
+
+    const offsets: PopperOffsets = { vertical: 20 };
+    return (
+        <div style={{ display: 'flex', justifyContent: 'center', height: 250, paddingTop: 100 }}>
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply demo/react/components/popover/realCase.tsx]: ok

File: src/components/popover/react/Popover.tsx [hunk 25]
```diff
@@ -0,0 +1,303 @@
...
+    useTooltipMode: false,
+};
+/////////////////////////////
+
+function unwrap(inputValue: boolean | string | (() => boolean) | undefined): boolean {
+    return typeof inputValue === 'function' ? inputValue() : Boolean(inputValue);
+}
+
+function computeOffsets(
+    placement: string,
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply src/components/popover/react/Popover.tsx]: ok

File: src/components/popover/react/Popover.tsx [hunk 26]
```diff
@@ -0,0 +1,303 @@
...
+ * @readonly
+ */
+const DEFAULT_PROPS: IDefaultPropsType = {
+    lockFlip: false,
+    tooltipShowHideDelay: SHOW_HIDE_DELAY,
+    useTooltipMode: false,
+};
+/////////////////////////////
+
+function unwrap(inputValue: boolean | string | (() => boolean) | undefined): boolean {
```
[the remark this answers is not in the document]
Comment (GuillaumeNachury) [reply src/components/popover/react/Popover.tsx]: ok

File: demo/react/components/popover/controled.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (Flo5k5) [reply demo/react/components/popover/controled.tsx]: okay!

**Comment by GuillaumeNachury** [comment]
I still don't get the point here and why it is blocking the merge.
The camel case is only used for demo samples and to be consistent (in the entry file where we define the sample) by not using the string representation with `-` . Since the property match the filename
```
{
    default: {
        //stuff
    },
    offsets: {
         //stuff
    },
    aDemoFileHyphens: {
       //stuff
    },
``` 
vs.
```
{
    default: {
        //stuff
    },
    offsets: {
         //stuff
    },
    'a-demo-file-hyphens': {
       //stuff
    },
```

**Pull request merged**